### ISSUE TITLE
Backend API specifications for M2

### ISSUE THREAD

**Issue by ersoykadir** [bounswe/bounswe2022group3#352]
### Issue

We need to decide on the necessary API specifications for the next milestone. The previous ones are [here](https://github.com/bounswe/bounswe2022group3/wiki/API-Specifications). Anyone from the front and mobile team can add related endpoints they need for the next milestone.

### Task(s)

- [ ] Update the old specifications with respect to implemented endpoints.
- [ ] Add new ones.

### Deliverable(s)

* Updated new specifications.

### Acceptance Criteria

* Additional new specifications commented under this issue and added to wiki.

### Deadline of the issue

11.11.2022 @23.59

**Comment by nurlandadashov02** [comment]
## Learning Spaces & Topics & Resources
### [POST] Create Topic:
#### Payload:
```
{
   course_id: "111111",
   topic: "Guitars"
}
```
#### Response: 
```
{ message: "" }
```
### [POST] Create Resource:
#### Payload:   (capture the author from token and current time for resource creation time, initial rating 0)
```
{
   course_id: "111111",
   topic_id: "222",
   title: "", 
   resource: "md-string"
}
```
#### Response: 
```
{ message: "" }
```
### [GET] Fetch learning space:
#### Query: `space_id`
#### Response: 
```
{ 
       title: "",
       owner: {
            name: "",
            surname: "",
            user_id: ""
      },
      rating: 4.5,
      topics: {
             topic_1_title: [
                  {
                        title: "resource_title_1",
                        resource_id: "111111"
                  },
                  {
                        title: "resource_title_2",
                        resource_id: "111111"
                  },
             ],
             topic_2_title: [
                  {
                        title: "resource_title_3",
                        resource_id: "111111"
                  },
                  {
                        title: "resource_title_4",
                        resource_id: "111111"
                  },
             ],
      }
}
```
### [GET] Fetch Resource:
#### Query: /[space_id]/[resource_id]
#### Response: 
```
{ 
        title: ""
        resource: "md-string",
        rating: 4.4, // average of ratings
        creator: {
            name: "",
            surname: "",
            user_id: ""
        },
        date_added: unix_timestamp (utc),
        discussion_id: "1222223feerrff"
}
```
### [POST] Edit Resource:
#### Payload:   (do we care who changed it and when?)
```
{
   course_id: "111111",
   topic_id: "222",
   resource_id: "1dss",
   title: "new_title", 
   resource: "new_md-string"
}
```
#### Response: 
```
{ message: "" }
```
### [POST] Rate Resource:
#### Payload:   (update average)
```
{
   course_id: "111111",
   topic_id: "222",
   resource_id: "1dss",
   rating: 2.3
}
```
#### Response: 
```
{ message: "" }
```
### [DELETE] Delete Resource:
#### Payload:   (check from token if the user is creator of resource or owner of workspace)
```
{
   course_id: "111111",
   topic_id: "222",
   resource_id: "1dss"
}
```
#### Response: 
```
{ message: "" }
```

**Comment by nurlandadashov02** [comment]
## Discussions
### [POST] Create Discussion:
#### Payload:
```
{
   course_id: "111111",
   title: "Guitars"
}
```
#### Response: 
```
{ message: "" }
```
### [POST] Create Comment in a discussion:
#### Payload:   (capture the author from token and current time for resource creation time, initial rating 0)
```
{
   course_id: "111111",
   discussion_id: "222",
   comment: "md-string"
}
```
#### Response: 
```
{ message: "" }
```
### [GET] Fetch discussion:
#### Query: `/course_id/discussion_id`
#### Response:  (only user who has joined a space can fetch discussion of a space)
```
{ 
       title: "",
       comments: [
              { // comment
                    author: {
                         name: "",
                         surname: "",
                         user_id: "",
                         image: "base-64 or url"
                    },
                    date_time: unix_timestamp, utc
                    text: "md-string"
              },
             { // comment
                    author: {
                         name: "",
                         surname: "",
                         user_id: "",
                         image: "base-64 or url"
                    },
                    date_time: unix_timestamp, utc
                    text: "md-string"
              },
      ]
}
```
### [GET] Fetch Discussion list:
#### Query: /[course_id]
#### Response: 
```
{ 
       discussions: [
             {
                     title: "",
                     discussion_id: ""
             },
             {
                     title: "",
                     discussion_id: ""
             },
       ]
}

**Comment by nurlandadashov02** [comment]
getEnrolledCourse - null, duplicate issue + keyword search

**Comment by nurlandadashov02** [comment]
update profile (name, surname, bio, interests)

**Comment by ersoykadir** [comment]
## User
### [POST] register:
#### Payload:
```
{
   email: "kadir.ersoy@boun.edu.tr",
   name: "kadir",
   surname: "ersoy",
   password: "Password123.",
   agreement: true
}
```
#### Response: 
```
{ message: "" }
```
### [POST] resend_confirmation:
#### Payload:
```
{
   email: "kadir.ersoy@boun.edu.tr",
}
```
#### Response: 
```
{ message: "" }
```

**Comment by furkanakkurt1335** [comment]
> getEnrolledCourse - null, duplicate issue + keyword search

Resolutions for both are in PRs #359 and #372 .

**Comment by salimtirit** [comment]
- Fetch learning space returns the topics with their resources but in the mobile app since the designs of the pages are different another API may be better in addition to it.
- On the web version the resources are shown under the topic titles
- On the mobile app the resources are shown when the user enters the topic. 


### [GET] Fetch Topic:
#### Query: /[space_id]/[topic_name]
#### Response: 
```
{ 
        resources: [  
                  {
                        title: "resource_title_3",
                        resource_id: "111111"
                  },
                  {
                        title: "resource_title_4",
                        resource_id: "111111"
                  },
       ]
}

**Comment by furkanakkurt1335** [comment]
## On [this comment](https://github.com/bounswe/bounswe2022group3/issues/352#issuecomment-1312186595)

- create topic and resource has been merged in #374 : `/topic`, `/resource`.
- get space detail (similar to 3rd) exists, adding resource titles to it; 1 endpoint: `/space/:id`.
- get resource: resource model didn't have creator. Adding it and creator info in the request data. [`/resource/:id`
- edit, rate, delete resource: done. [`/resource/update`, `resource/rate`, `resource/delete`]

## On [this comment](https://github.com/bounswe/bounswe2022group3/issues/352#issuecomment-1327644067)

- get topic: done, `/topic/mini/:id`.

**Comment by hsuleerkul** [comment]
## Profile Page

### [POST] Edit Profile:

Payload:

```
{
 user_id: "",
 email: "",
 name: "",
 surname: "",
 followed_users: [],
 follower_users: [],
 blocked_users: [],
 is_private: "",
 personal_info: [
			id: "",
			bio: "",
			interests: [],
			knowledge: [],
 		       ]
}
```

Response:

```
{ message: "" }

```

### [GET] Get Profile:

Query: `user_id`

Response: 

```
{
user_id: "",
email: "",
name: "",
surname: "",
enrollments: [],
created_spaces: [],
followed_users: [],
follower_users: [],
blocked_users: [],
is_private: "",
personal_info: [
			id: "",
			bio: "",
			personal_achievements: [],
			personal_activities: []
			interests: [],
			knowledge: [],
			badges: []
 		      ]
}
```
